haskell-src/Concordium/Types/ProtocolVersion.hs

```diff
@@ -20,38 +32,253 @@
+    -- to the hashing scheme.
+    --
+    --   * 'TOVO' is used in P1 to P4. The hash is computed as a simple hash list.
+    --     All the contents of the transaction summaries are used for computing the hash.
+    --
+    --   * 'TOV1' is used in PV5 and onwards. The hash is computed via a merkle tree and the
+    --     exact reject reasons for failed transactions are omitted from the hash.
+    TransactionOutcomesVersion (..),
+    -- |Singleton type corresponding to 'TransactionOutcomesVersion'.
+    STransactionOutcomesVersion (..),
+    IsTransactionOutcomesVersion,
+    transactionOutcomesVersion,
+    -- |The transaction outcomes version for a given protocol version.
+    transactionOutcomesVersionFor,
+    -- |The transaction outcomes version for a given protocol version (at the type level).
+    TransactionOutcomesVersionFor,
+    -- |The transaction outcomes version for a given protocol version (on singletons).
+    sTransactionOutcomesVersionFor,
+
+    -- * Delegation support
+
+    -- |Whether or not delegation is supported at a particular account version.
+    DelegationSupport (..),
+    -- Singleton type corresponding to 'DelegationSupport'.
+    SDelegationSupport (..),
+    -- |Determine whether delegation is supported for a particular account version.
+    supportsDelegation,
+    -- |Determine whether delegation is supported for a particular account version (at the type level).
+    SupportsDelegation,
+    -- |Determine whether delegation is supported for a particular account version (on singletons).
+    sSupportsDelegation,
+    SAVDelegationSupport (..),
+    AVSupportsDelegation,
+    PVSupportsDelegation,
+    delegationSupport,
+    protocolSupportsDelegation,
+
+    -- * Feature guards
+    supportsMemo,
+    supportsV1Contracts,
+    supportsUpgradableContracts,
+    supportsChainQueryContracts,
+    supportsAccountAliases,
+    supportsDelegationPV,
+
+    -- * Defunctionalisation symbols
+    P1Sym0,
+    P2Sym0,
+    P3Sym0,
+    P4Sym0,
+    P5Sym0,
+    P6Sym0,
+) where
 
 import Data.Aeson
 import Data.Aeson.Types
 import Data.Serialize
-import Data.Word
+import Data.Singletons.Base.TH
 
--- |An enumeration of the supported versions of the consensus protocol.
--- Binary and JSON serializations are as Word64 corresponding to the protocol number.
-data ProtocolVersion
-    = P1
-    | P2
-    | P3
-    | P4
-    | P5
-    | P6
-    deriving (Eq, Show, Ord)
-
--- |The singleton type associated with 'ProtocolVersion'.
--- There is a unique constructor of 'SProtocolVersion' for
--- each constructor of 'ProtocolVersion'.
-data SProtocolVersion (pv :: ProtocolVersion) where
-    SP1 :: SProtocolVersion 'P1
-    SP2 :: SProtocolVersion 'P2
-    SP3 :: SProtocolVersion 'P3
-    SP4 :: SProtocolVersion 'P4
-    SP5 :: SProtocolVersion 'P5
-    SP6 :: SProtocolVersion 'P6
+import Data.Word
 
+$( singletons
+    [d|
+        -- \|An enumeration of the supported versions of the consensus protocol.
```

Comment:
One cannot generate haddoc documentation within template haskell, so that is why it's in the header and export list. 
The reason for having it here in the template haskell part is to also provide doc when e.g. jumping to definition.

I admit it's a bit duplication, but I think it's worth it?

Reply:
The reason for having it here is more so you can see what it's meant to do when you're reading the code, rather than having to look elsewhere.